http-server/src/server.rs

```diff
@@ -384,9 +439,185 @@ impl Future for ServerHandle {
+					None => return Ok(malformed()),
+				};
+
+				let allowed_headers = acl.allowed_headers().to_cors_header_value();
+				let allowed_header_bytes = allowed_headers.as_bytes();
+
+				let res = hyper::Response::builder()
+					.header("access-control-allow-origin", origin)
+					.header("access-control-allow-methods", "POST")
+					.header("access-control-allow-headers", allowed_header_bytes)
+					.body(hyper::Body::empty())
+					.unwrap_or_else(|e| {
+						tracing::error!("Error forming preflight response: {}", e);
+						internal_error()
+					});
+
+				Ok(res)
+			}
+			// The actual request. If it's a CORS request we need to remember to add
+			// the access-control-allow-origin header (despite preflight) to allow it
+			// to be read in a browser.
+			Method::POST if content_type_is_json(&request) => {
+				let origin = return_origin_if_different_from_host(request.headers()).cloned();
+				let mut res = process_validated_request(ProcessValidatedRequest {
+					request,
+					logger,
+					methods,
+					resources,
+					max_request_body_size,
+					max_response_body_size,
+					max_log_length,
+					batch_requests_supported,
+					request_start,
+				})
+				.await?;
+
+				if let Some(origin) = origin {
+					res.headers_mut().insert("access-control-allow-origin", origin);
+				}
+				Ok(res)
+			}
+			Method::GET => match health_api.as_ref() {
+				Some(health) if health.path.as_str() == request.uri().path() => {
+					process_health_request(
+						health,
+						logger,
+						methods,
+						max_response_body_size,
+						request_start,
+						max_log_length,
+					)
+					.await
+				}
+				_ => Ok(response::method_not_allowed()),
+			},
+			// Error scenarios:
+			Method::POST => Ok(response::unsupported_content_type()),
+			_ => Ok(response::method_not_allowed()),
+		}
+	}
+}
+
+/// JsonRPSee service compatible with `tower`.
+///
+/// # Note
+/// This is similar to [`hyper::service::service_fn`].
+#[derive(Debug)]
+pub struct TowerService<L> {
+	inner: ServiceData<L>,
+}
+
+impl<L: Logger> hyper::service::Service<hyper::Request<hyper::Body>> for TowerService<L> {
+	type Response = hyper::Response<hyper::Body>;
+
+	// NOTE(lexnv): The `handle_request` method returns `Result<_, Infallible>`.
+	// This is because the RPC service will always return a valid HTTP response (ie return `Ok(_)`).
+	//
+	// The following associated type is required by the `impl<B, U, L: Logger> Server<B, L>` bounds.
+	// It satisfies the server's bounds when the `tower::ServiceBuilder<B>` is not set (ie `B: Identity`).
+	type Error = Box<dyn StdError + Send + Sync + 'static>;
+
+	type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send>>;
+
+	/// Opens door for back pressure implementation.
+	fn poll_ready(&mut self, _: &mut Context) -> Poll<Result<(), Self::Error>> {
+		Poll::Ready(Ok(()))
+	}
+
+	fn call(&mut self, request: hyper::Request<hyper::Body>) -> Self::Future {
+		let data = self.inner.clone();
```

Comment:
Do we have to worry about performance at all cloning all of these?

(It looked like a bunch of them were cloned previously so I'm guessing not).

And ah ok, is the inner struct so that you can clone it here? I guess you can just clone `&mut self` though and have just one struct?

Reply:
I think it's similar as it was before because of the `FnMut` closure in the hyper service fn